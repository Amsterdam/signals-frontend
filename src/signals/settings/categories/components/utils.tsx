// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
import type { FieldNamesMarkedBoolean } from 'react-hook-form'

import type { StandardText } from 'types/api/standard-texts'

import type { CategoryFormPayload, CategoryFormValues } from '../types'
import { Direction } from '../types'

export const getPatchPayload = (
  formData: CategoryFormValues,
  dirtyFields: Partial<Readonly<FieldNamesMarkedBoolean<CategoryFormValues>>>
): CategoryFormPayload => {
  const payload = Object.keys(dirtyFields).map((key) => {
    return {
      [key]: formData[key as keyof CategoryFormValues],
    }
  })

  const payloadObject = Object.assign({}, ...payload)

  const payloadResult = {
    ...payloadObject,
    ...(('n_days' in payloadObject || 'use_calendar_days' in payloadObject) && {
      new_sla: {
        n_days: payloadObject.n_days ?? formData.n_days,
        use_calendar_days: Boolean(Number(payloadObject.use_calendar_days)),
      },
    }),
    ...('show_children_in_filter' in payloadObject && {
      configuration: {
        show_children_in_filter: payloadObject.show_children_in_filter,
      },
    }),
  }

  delete payloadResult['n_days']
  delete payloadResult['use_calendar_days']
  delete payloadResult['show_children_in_filter']
  delete payloadResult['standard_texts']

  return payloadResult
}

export const orderStandardTexts = (
  direction: Direction,
  index: number,
  orderedStandardTexts: StandardText[]
): StandardText[] | undefined => {
  const orderedStandardTextsCopy = [...orderedStandardTexts]
  if (direction === Direction.Up) {
    if (index === 0) {
      return
    }
    const temp = orderedStandardTextsCopy[index - 1]
    orderedStandardTextsCopy[index - 1] = orderedStandardTextsCopy[index]
    orderedStandardTextsCopy[index] = temp
  }
  if (direction === Direction.Down) {
    if (index === orderedStandardTextsCopy.length - 1) {
      return
    }
    const temp = orderedStandardTextsCopy[index + 1]
    orderedStandardTextsCopy[index + 1] = orderedStandardTextsCopy[index]
    orderedStandardTextsCopy[index] = temp
  }
  return orderedStandardTextsCopy
}
