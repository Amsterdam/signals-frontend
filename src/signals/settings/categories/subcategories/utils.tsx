// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
import type {
  CategoryFormValues,
  CategoryFormPayload,
  DirtyFields,
} from '../types'

export const getPatchPayload = (
  formData: CategoryFormValues,
  dirtyFields: DirtyFields
): CategoryFormPayload => {
  const payload = Object.keys(dirtyFields).map((key) => {
    return {
      [key]: formData[key as keyof CategoryFormValues],
    }
  })

  const payloadObject = Object.assign({}, ...payload)

  if (payloadObject['n_days'] || payloadObject['use_calendar_days']) {
    const updatedPayload = {
      ...payloadObject,
      new_sla: {
        n_days: payloadObject.n_days ?? formData.n_days,
        use_calendar_days: Boolean(Number(payloadObject.use_calendar_days)),
      },
    }

    delete updatedPayload['n_days']
    delete updatedPayload['use_calendar_days']

    return updatedPayload
  } else return payloadObject
}
