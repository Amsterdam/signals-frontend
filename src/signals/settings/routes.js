export const BASE_URL = '/instellingen';
export const USERS_URL = `${BASE_URL}/gebruikers`;
export const USERS_PAGED_URL = `${USERS_URL}/page`;
export const USER_URL = `${BASE_URL}/gebruiker`;
export const ROLES_URL = `${BASE_URL}/rollen`;
export const ROLE_URL = `${BASE_URL}/rol`;
export const DEPARTMENTS_URL = `${BASE_URL}/afdelingen`;
export const DEPARTMENT_URL = `${BASE_URL}/afdeling`;
export const CATEGORY_URL = `${BASE_URL}/categorie`;
export const CATEGORIES_URL = `${BASE_URL}/categorieen`;
export const CATEGORIES_PAGED_URL = `${CATEGORIES_URL}/page`;

const routes = {
  users: USERS_URL,
  usersPaged: `${USERS_PAGED_URL}/:pageNum(\\d+)`,
  user: `${USER_URL}/:userId(\\d+)`,

  roles: ROLES_URL,
  role: `${ROLE_URL}/:roleId(\\d+)`,

  departments: DEPARTMENTS_URL,
  department: `${DEPARTMENT_URL}/:departmentId(\\d+)`,

  categories: CATEGORIES_URL,
  categoriesPaged: `${CATEGORIES_PAGED_URL}/:pageNum(\\d+)`,
  category: `${CATEGORY_URL}/:categoryId(\\d+)`,
};

export default routes;
