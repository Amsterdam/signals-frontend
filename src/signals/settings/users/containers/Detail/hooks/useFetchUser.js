import { useState, useEffect } from 'react';

import CONFIGURATION from 'shared/services/configuration/configuration';
import { getAuthHeaders } from 'shared/services/auth/auth';

export const userEndpoint = `${CONFIGURATION.API_ROOT}signals/v1/private/users/`;

/**
 * Custom hook useFetchUser
 *
 * Will call private /users endpoint
 *
 * @returns {FetchResponse}
 */
const useFetchUser = id => {
  const [isLoading, setLoading] = useState();
  const [data, setData] = useState();
  const [error, setError] = useState(false);

  useEffect(() => {
    const controller = new AbortController();
    const { signal } = controller;

    async function fetchData() {
      setLoading(true);

      try {
        const url = [userEndpoint, id].join('');
        const response = await fetch(url, {
          headers: getAuthHeaders(),
          signal,
        });
        const userData = await response.json();

        setData(userData);
      } catch (e) {
        setError(e);
      } finally {
        setLoading(false);
      }
    }

    fetchData();

    return (() => {
      controller.abort();
    });
  }, []);

  /**
   * @typedef {Object} FetchResponse
   * @property {Boolean} isLoading - Indicator of fetch state
   * @property {Object} data - User object
   * @property {Error} error - Error object thrown during fetch and data parsing
   */
  return { isLoading, data, error };
};

export default useFetchUser;
