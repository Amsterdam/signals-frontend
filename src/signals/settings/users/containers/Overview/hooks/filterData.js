// name mapping from API values to human readable values
export const colMap = {
  id: 'id',
  is_active: 'Status',
  roles: 'Rol',
  username: 'Gebruikersnaam',
};

/**
 * Clean-up of users API response
 *
 * Filtering out invalid keys, turning array values into concatenated strings and
 * converting boolean values to readable text values
 *
 * @param {Object} data
 * @returns {Object}
 */
const filterData = data => {
  const allowedKeys = Object.keys(colMap);

  return data.map(item =>
    Object.keys(item)
      .filter(key => allowedKeys.includes(key))
      .reduce((rawObj, key) => {
        const obj = { ...rawObj };
        let value = Array.isArray(item[key])
          ? item[key].join(', ')
          : item[key];

        if (typeof value === 'boolean') {
          value = value ? 'Actief' : 'Niet actief';
        }

        obj[colMap[key]] = value;

        return obj;
      }, {})
  );
};

export default filterData;
