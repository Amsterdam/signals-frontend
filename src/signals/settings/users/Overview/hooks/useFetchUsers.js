import { useState, useEffect } from 'react';

import { PAGE_SIZE } from 'containers/App/constants';
import { getAuthHeaders } from 'shared/services/auth/auth';
import configuration from 'shared/services/configuration/configuration';

import filterData from '../../../filterData';

// name mapping from API values to human readable values
const colMap = {
  id: 'id',
  is_active: 'Status',
  roles: 'Rol',
  username: 'Gebruikersnaam',
};

/**
 * Custom hook useFetchUsers
 *
 * Will call private /users endpoint
 *
 * @returns {FetchResponse}
 */
const useFetchUsers = ({ page, filters } = {}) => {
  const [isLoading, setLoading] = useState(false);
  const [users, setUsers] = useState([]);
  const [errorState, setError] = useState(false);

  useEffect(() => {
    const controller = new AbortController();
    const { signal } = controller;

    (async function fetchData() {
      setLoading(true);

      const pageParams = [page && `page=${page}`, `page_size=${PAGE_SIZE}`];

      const filterParams = Object.entries(filters || {})
        .filter(([, value]) => value !== '*')
        .reduce((acc, [filter, value]) => [...acc, `${filter}=${value}`], []);

      const queryParams = [...pageParams, ...filterParams]
        .filter(Boolean)
        .join('&');

      try {
        const url = [configuration.USERS_ENDPOINT, queryParams].join('?');

        const response = await fetch(url, {
          headers: {
            ...getAuthHeaders(),
            Accept: 'application/json',
          },
          signal,
        });

        const userData = await response.json();
        const filteredUserData = filterData(userData.results, colMap);

        setUsers({ count: userData.count, list: filteredUserData });
      } catch (error) {
        setError(error);
      } finally {
        setLoading(false);
      }
    })();

    return () => {
      controller.abort();
    };
  }, [page, filters]);

  /**
   * @typedef {Object} FetchResponse
   * @property {Boolean} isLoading - Indicator of fetch state
   * @property {Object[]} users - Array of user objects
   * @property {Error} error - Error object thrown during fetch and data parsing
   */
  return { isLoading, users, error: errorState };
};

export default useFetchUsers;
