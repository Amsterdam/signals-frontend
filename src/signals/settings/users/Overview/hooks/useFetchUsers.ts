// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2019 - 2021 Gemeente Amsterdam
import { useState, useEffect, useCallback } from 'react'

import { PAGE_SIZE as page_size } from 'containers/App/constants'
import useFetch from 'hooks/useFetch'
import type { FetchError } from 'hooks/useFetch'
import configuration from 'shared/services/configuration/configuration'
import type UsersData from 'types/api/users'

import filterData from '../../../utils/filterData'

export type Filters = {
  profile_department_code?: string
  role?: string
  is_active?: 'true' | 'false'
  username?: string
}

// name mapping from API values to human readable values
const colMap = {
  id: 'id',
  is_active: 'Status',
  roles: 'Rol',
  username: 'Gebruikersnaam',
  profile: 'profile',
}

type FetchUsersProps = {
  filters?: URLSearchParams
  page?: number
}

type User = {
  id: number
  Status: boolean
  Rol: Array<string>
  Gebruikersnaam: string
  Afdeling: string
}

type Users = {
  count: number
  list: Array<User>
}

interface FetchResponse {
  isLoading: boolean
  users: Users
  error?: boolean | FetchError
  get: (params?: FetchUsersProps) => void
}

const useFetchUsers = (): FetchResponse => {
  const [users, setUsers] = useState<Users>({ count: 0, list: [] })
  const { get: fetch, data, isLoading, error } = useFetch<UsersData>()

  const get = useCallback(
    async ({ filters, page }: FetchUsersProps = {}) => {
      const pageParams = { page, page_size }
      const entries = filters ? Array.from(filters.entries()) : []

      const filterParams = entries.length
        ? entries
            .filter(
              ([key, value]: [string, string]) => value !== '*' && key !== ''
            )
            .reduce(
              (
                acc: Record<string, string>,
                [filter, value]: [string, string]
              ) => ({ ...acc, [filter]: value }),
              {}
            )
        : {}

      const queryParams = { ...pageParams, ...filterParams }

      await fetch(configuration.USERS_ENDPOINT, queryParams)
    },
    [fetch]
  )

  useEffect(() => {
    if (!data) return

    const filteredUserData: Array<User> = filterData(data.results, colMap).map(
      ({ profile, id, Status, Rol, Gebruikersnaam }) => ({
        id,
        Status,
        Rol,
        Gebruikersnaam,
        Afdeling: profile?.departments ? profile.departments.join(', ') : '',
      })
    )

    setUsers({ count: data.count, list: filteredUserData })
  }, [data])

  return { get, users, isLoading, error }
}

export default useFetchUsers
