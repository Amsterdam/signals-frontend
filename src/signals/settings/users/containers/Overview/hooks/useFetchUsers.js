import { useState, useEffect } from 'react';

import CONFIGURATION from 'shared/services/configuration/configuration';
import { getAuthHeaders } from 'shared/services/auth/auth';

import filterData from './filterData';

export const usersEndpoint = `${CONFIGURATION.API_ROOT}signals/v1/private/users`;

/**
 * Custom hook useFetchUsers
 *
 * Will call private /users endpoint
 *
 * @returns {FetchResponse}
 */
const useFetchUsers = ({ page, pageSize } = {}) => {
  const [isLoading, setLoading] = useState(false);
  const [users, setUsers] = useState([]);
  const [error, setError] = useState(false);

  useEffect(() => {
    const controller = new AbortController();
    const { signal } = controller;

    async function fetchData() {
      setLoading(true);

      try {
        const params = [
          page && `page=${page}`,
          pageSize && `page_size=${pageSize}`,
        ]
          .filter(Boolean)
          .join('&');
        const url = [usersEndpoint, params].filter(Boolean).join('/?');
        const response = await fetch(url, {
          headers: getAuthHeaders(),
          signal,
        });
        const userData = await response.json();
        const filteredUserData = filterData(userData.results);

        setUsers({ count: userData.count, list: filteredUserData });
      } catch (e) {
        setError(e);
      } finally {
        setLoading(false);
      }
    }

    fetchData();

    return () => {
      controller.abort();
    };
  }, [page]);

  /**
   * @typedef {Object} FetchResponse
   * @property {Boolean} isLoading - Indicator of fetch state
   * @property {Object[]} users - Array of user objects
   * @property {Error} error - Error object thrown during fetch and data parsing
   */
  return { isLoading, users, error };
};

export default useFetchUsers;
