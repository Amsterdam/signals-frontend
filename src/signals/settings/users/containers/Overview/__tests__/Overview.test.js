import React from 'react';
import { act } from '@testing-library/react-hooks';
import { render, wait, fireEvent, cleanup } from '@testing-library/react';
import { act as reAct } from 'react-dom/test-utils';
import * as reactRouterDom from 'react-router-dom';
import { withAppContext, history } from 'test/utils';
import usersJSON from 'utils/__tests__/fixtures/users.json';
import routes from 'signals/settings/routes';

import UsersOverview from '..';
import { usersEndpoint } from '../hooks/useFetchUsers';

jest.mock('react-router-dom', () => {
  const actual = jest.requireActual('react-router-dom');
  return {
    __esModule: true,
    ...actual,
    useParams: jest.fn(() => ({})),
    useLocation: () => ({
      hash: '',
      key: '',
      pathname: '/instellingen/gebruikers/',
      search: '',
      state: null,
    }),
  };
});

describe('signals/settings/users/containers/Overview', () => {
  beforeAll(() => {
    fetch.mockResponse(JSON.stringify(usersJSON));
  });

  const historyMock = {
    ...history,
    action: '',
  };

  it('should request users from API on mount', async () => {
    await act(async () => {
      await render(withAppContext(<UsersOverview history={historyMock} />));

      await wait(() =>
        expect(global.fetch).toHaveBeenCalledWith(
          expect.stringContaining(usersEndpoint),
          expect.objectContaining({ headers: {} })
        )
      );
    });
  });

  it('should render the list of fetched users', async () => {
    let getByText;
    let container;

    await reAct(async () => {
      ({ container, getByText } = await render(
        withAppContext(<UsersOverview history={historyMock} />)
      ));
    });

    expect(getByText('Gebruikers (69)')).toBeTruthy();

    expect(container.querySelectorAll('tbody tr')).toHaveLength(69);
  });

  it('should render pagination controls', async () => {
    let queryByTestId;
    let rerender;

    await reAct(async () => {
      ({ rerender, queryByTestId } = await render(
        withAppContext(<UsersOverview history={historyMock} />)
      ));
    });

    expect(queryByTestId('overviewPagerComponent')).toBeTruthy();

    cleanup();

    await reAct(async () => {
      await rerender(
        withAppContext(<UsersOverview pageSize={100} history={historyMock} />)
      );
    });

    expect(queryByTestId('overviewPagerComponent')).toBeFalsy();
  });

  it('should push to the history stack on pager item click', async () => {
    let container;
    const historyMockObj = { ...historyMock, push: jest.fn() };

    await reAct(async () => {
      ({ container } = await render(
        withAppContext(<UsersOverview history={historyMockObj} />)
      ));
    });

    fireEvent.click(container.querySelector('.pager_nav.volgende'));
    expect(historyMockObj.push).toHaveBeenCalledWith(
      expect.stringContaining(routes.users)
    );
  });

  it('should push on update when page parameter and page state var differ', async () => {
    const historyMockObj = {
      ...historyMock,
      action: 'PUSH',
      push: jest.fn(),
      replace: jest.fn(),
    };

    jest.spyOn(reactRouterDom, 'useParams').mockImplementation(() => ({
      pageNum: 2,
    }));

    await reAct(async () => {
      await render(withAppContext(<UsersOverview history={historyMockObj} />));
    });

    expect(historyMockObj.push).toHaveBeenCalledWith(
      expect.stringContaining(`${routes.users}/page/1`)
    );

    expect(historyMockObj.replace).not.toHaveBeenCalled();
  });

  it('should not push', async () => {
    const historyMockObj = {
      ...historyMock,
      action: 'PUSH',
      push: jest.fn(),
    };

    jest.spyOn(reactRouterDom, 'useParams').mockImplementation(() => ({
      pageNum: 1,
    }));

    await reAct(async () => {
      await render(withAppContext(<UsersOverview history={historyMockObj} />));
    });

    expect(historyMockObj.push).not.toHaveBeenCalled();
  });

  it('should not push on POP', async () => {
    const historyMockObj = {
      ...historyMock,
      action: 'POP',
      push: jest.fn(),
    };

    jest.spyOn(reactRouterDom, 'useParams').mockImplementation(() => ({
      pageNum: 1,
    }));

    await reAct(async () => {
      await render(withAppContext(<UsersOverview history={historyMockObj} />));
    });

    expect(historyMockObj.push).not.toHaveBeenCalled();
  });

  it('should push on list item click', async () => {
    let container;
    const historyMockObj = {
      ...historyMock,
      push: jest.fn(),
    };

    await reAct(async () => {
      ({ container } = await render(
        withAppContext(<UsersOverview history={historyMockObj} />)
      ));
    });

    const row = container.querySelector('tbody tr:nth-child(42)');
    const itemId = 666;
    row.dataset.itemId = itemId;

    fireEvent.click(row.querySelector('td:first-of-type'));

    expect(historyMockObj.push).toHaveBeenCalledWith(
      routes.user.replace(':userId', itemId)
    );

    historyMockObj.push.mockReset();

    const row2 = container.querySelector('tbody tr:nth-child(43)');
    delete row2.dataset.itemId;

    fireEvent.click(row2.querySelector('td:first-of-type'));

    expect(historyMockObj.push).not.toHaveBeenCalled();
  });
});
