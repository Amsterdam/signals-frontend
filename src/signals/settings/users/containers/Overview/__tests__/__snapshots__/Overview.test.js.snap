// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signals/settings/users/containers/Overview should push to the history stack on pagination item click 1`] = `
<section
  class="PageHeader__StyledSection-zrug1o-0 diDgeL PageHeader__StyledPageHeader-sc-1via2ti-0 dsIYGM"
>
  <div
    class="sc-kgAjT iCvEVS"
  >
    <div>
      <h1
        class="sc-EHOje jjCRJL sc-bZQynM hJEdCc PageHeader__StyledHeading-zrug1o-1 gmiYbT"
      >
        Gebruikers (69)
      </h1>
      
    </div>
  </div>
</section>
`;
